File: spring-jpa/boot-jpa-if/src/main/java/jpabook/jpashop/service/OrderService.java
package jpabook.jpashop.service;

import jpabook.jpashop.domain.Delivery;
import jpabook.jpashop.domain.Member;
import jpabook.jpashop.domain.Order;
import jpabook.jpashop.domain.OrderItem;
import jpabook.jpashop.domain.item.Item;
import jpabook.jpashop.repository.ItemRepository;
import jpabook.jpashop.repository.MemberRepository;
import jpabook.jpashop.repository.OrderRepository;
import jpabook.jpashop.repository.OrderSearch;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional(readOnly = true)
public class OrderService {

    private final OrderRepository orderRepository;
    private final MemberRepository memberRepository;
    private final ItemRepository itemRepository;

    public OrderService(OrderRepository orderRepository, MemberRepository memberRepository, ItemRepository itemRepository) {
        this.orderRepository = orderRepository;
        this.memberRepository = memberRepository;
        this.itemRepository = itemRepository;
    }


    // 주문
    @Transactional
    public Long order(Long memberId, Long itemId, int count) {

        // 엔티티 조회
        Member member = memberRepository.findById(memberId).orElseThrow();
        Item item = itemRepository.findOne(itemId);

        // 배송정보 생성
        Delivery delivery = new Delivery();
        delivery.setAddress(member.getAddress());

        // 주문상품 생성
        OrderItem orderItem = OrderItem.createOrderItem(item, item.getPrice(), count);

        // 주문 생성
        Order order = Order.createOrder(member, delivery, orderItem);

        orderRepository.save(order);

        return order.getId();
    }


    // 취소
    @Transactional
    public void cancelOrder(Long orderId) {
        Order order = orderRepository.findOne(orderId);
        order.cancel();
    }

    // 검색
    public List<Order> findOrders(OrderSearch orderSearch) {
        return orderRepository.findAll(orderSearch);
    }

    public List<Order> findAndWithMemberDelivery() {
        return orderRepository.findAndWithMemberDelivery();
    }
}

File: architecture-usecase-1/src/main/kotlin/org/example/usecase/service/OrderService.kt
package org.example.usecase.service

import org.example.usecase.dto.NewOrder
import org.example.usecase.dto.OrderDto
import org.example.usecase.entity.OrderEntity
import org.example.usecase.repository.OrderRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class OrderService(
    private val orderRepository: OrderRepository,
) {
    fun getOrder(id: Long): OrderDto = orderRepository
        .findByIdOrNull(id)
        ?.let {
            OrderDto(
                id = it.id,
                status = it.status,
            )
        }
        ?: throw IllegalArgumentException("Order not found")

    @Transactional
    fun createOrder(newOrder: NewOrder): Long {
        val order = orderRepository.save(
            OrderEntity(
                status = newOrder.status,
            )
        )

        return order.id
    }

}
